File: scripts/peass.py
import os
import musdb
import soundfile as sf
import numpy as np
import concurrent.futures
import uuid
import shutil
import multiprocessing
from pyass.peass import PEASS_ObjectiveMeasure

def process_single_chunk(start, end, reference_files, estimate_file, sr, peass_options):
    uid = str(uuid.uuid4())
    
    # Create a unique destDir for this chunk to avoid internal temp file collisions
    chunk_peass_options = peass_options.copy()
    chunk_dest_dir = os.path.join(peass_options['destDir'], uid)
    os.makedirs(chunk_dest_dir, exist_ok=True)
    chunk_peass_options['destDir'] = chunk_dest_dir
    
    # Read all arrays first to find the minimum length
    ref_arrays = []
    for i, ref_file in enumerate(reference_files):
        data, _ = sf.read(ref_file, start=start, frames=(end-start))
        ref_arrays.append(data)
        
    est_chunk, _ = sf.read(estimate_file, start=start, frames=(end-start))
    
    # Truncate to the shortest array to avoid broadcast errors
    min_len = min([len(est_chunk)] + [len(r) for r in ref_arrays])
    
    temp_refs = []
    for i, data in enumerate(ref_arrays):
        truncated_data = data[:min_len]
        temp_ref_path = f".temp_chunk_ref_{i}_{uid}.wav"
        sf.write(temp_ref_path, truncated_data, sr)
        temp_refs.append(temp_ref_path)
    
    est_chunk = est_chunk[:min_len]
    temp_est_path = f".temp_chunk_est_{uid}.wav"
    sf.write(temp_est_path, est_chunk, sr)
    
    # Calculate PEASS for the chunk
    chunk_scores = PEASS_ObjectiveMeasure(temp_refs, temp_est_path, chunk_peass_options)
    
    # Cleanup temp files and the unique PEASS destDir
    for path in temp_refs + [temp_est_path]:
        if os.path.exists(path):
            os.remove(path)
    shutil.rmtree(chunk_dest_dir, ignore_errors=True)
    
    return chunk_scores


def compute_peass_in_chunks(reference_files, estimate_file, peass_options, chunk_duration_sec=10):
    # Read the estimate file to get the samplerate and total length
    # Note: Using info to avoid loading the whole file if possible
    info = sf.info(estimate_file)
    sr = info.samplerate
    total_frames = info.frames
    chunk_frames = int(chunk_duration_sec * sr)
    
    ops_scores, tps_scores, ips_scores, aps_scores = [], [], [], []
    
    tasks = []
    # Collect chunk coordinates
    for start in range(0, total_frames, chunk_frames):
        end = min(start + chunk_frames, total_frames)
        if end - start == 0:
            break
        tasks.append((start, end))
            
    total_chunks = len(tasks)
    # Reduced to 5 to avoid Out-Of-Memory (OOM) stalls. PEASS is highly memory-intensive,
    # and running 10 parallel instances likely exhausts system RAM, causing severe disk swapping.
    max_workers = min(5, multiprocessing.cpu_count()) 
    print(f"    -> Dispatching {total_chunks} chunks to {max_workers} parallel workers...")
    
    with concurrent.futures.ProcessPoolExecutor(max_workers=max_workers) as executor:
        futures = [
            executor.submit(process_single_chunk, start, end, reference_files, estimate_file, sr, peass_options)
            for start, end in tasks
        ]
        
        for i, future in enumerate(concurrent.futures.as_completed(futures)):
            print(f"    -> Completed chunk {i + 1}/{total_chunks}...")
            chunk_scores = future.result()
            ops_scores.append(chunk_scores['OPS'])
            tps_scores.append(chunk_scores['TPS'])
            ips_scores.append(chunk_scores['IPS'])
            aps_scores.append(chunk_scores['APS'])

    # Return averaged scores
    return {
        'OPS': np.mean(ops_scores) if ops_scores else 0,
        'TPS': np.mean(tps_scores) if tps_scores else 0,
        'IPS': np.mean(ips_scores) if ips_scores else 0,
        'APS': np.mean(aps_scores) if aps_scores else 0
    }

def main():
    # ==========================================
    # 1. CONFIGURATION
    # ==========================================
    mp4_dataset_path = "data/musdb18" 
    wav_dataset_path = ".temp_wavs"
    
    # Path to your model's predictions (e.g., Spleeter outputs)
    estimates_base_path = "output" # TODO: Compute Output
    
    # The stem you want to evaluate
    target_stem = "vocals" 
    
    # All possible MUSDB stems
    all_stems = ["vocals", "drums", "bass", "other"]
    
    # Directory for PEASS temporary calculation files
    peass_options = {'destDir': './peass_temp_output/'}
    os.makedirs(peass_options['destDir'], exist_ok=True)

    # ==========================================
    # 2. INITIALIZE DATASET
    # ==========================================
    print("Loading MP4 dataset...")
    mus = musdb.DB(root=mp4_dataset_path, is_wav=False, subsets="test")
    print(f"Found {len(mus)} tracks in the test subset.")
    
    # ==========================================
    # 3. DECODE & EVALUATE LOOP
    # ==========================================
    for track in mus:
        print(f"\nProcessing: {track.name}")
        
        # --- Decode Step ---
        track_wav_dir = os.path.join(wav_dataset_path, track.subset, track.name)
        os.makedirs(track_wav_dir, exist_ok=True)
        
        target_wav_path = os.path.join(track_wav_dir, f"{target_stem}.wav")
        
        # Only decode if the WAV files don't already exist from a previous run
        if not os.path.exists(target_wav_path):
            print("  Decoding MP4 to WAV...")
            # Save mixture
            sf.write(os.path.join(track_wav_dir, 'mixture.wav'), track.audio, track.rate)
            # Save stems
            for stem_name, stem_data in track.targets.items():
                sf.write(os.path.join(track_wav_dir, f'{stem_name}.wav'), stem_data.audio, track.rate)
        else:
            print("  WAV files already exist. Skipping decode.")

        # --- Evaluate Step ---
        # Locate your model's estimate for this specific track
        estimate_file = os.path.join(estimates_base_path, track.name + '.stem', f"{target_stem}.wav")
        
        if not os.path.exists(estimate_file):
            print(f"  [Warning] Estimate not found at {estimate_file}. Skipping evaluation.")
            continue

        # Construct the reference list (Target MUST be first)
        reference_files = [target_wav_path]
        
        # Add the remaining stems as interferences
        for stem in all_stems:
            if stem != target_stem:
                interf_path = os.path.join(track_wav_dir, f"{stem}.wav")
                reference_files.append(interf_path)

        print("  Computing PEASS metrics in chunks...")
        try:
            peass_scores = compute_peass_in_chunks(reference_files, estimate_file, peass_options, chunk_duration_sec=10)
            
            print(f"  -> OPS (Overall):      {peass_scores['OPS']:.2f}")
            print(f"  -> TPS (Target):       {peass_scores['TPS']:.2f}")
            print(f"  -> IPS (Interference): {peass_scores['IPS']:.2f}")
            print(f"  -> APS (Artifacts):    {peass_scores['APS']:.2f}")
            
        except Exception as e:
            print(f"  [Error] Failed to evaluate {track.name}: {e}")


if __name__ == "__main__":
    main()

File: scripts/peass.m
% ==========================================
% 1. CONFIGURATION
% ==========================================
addpath(genpath('scripts/peass-software'));

% NOTE: This assumes MUSDB18 tracks are already extracted to WAV format 
% in wav_dataset_path, as MATLAB doesn't have a native MUSDB mp4 decoder.

wav_dataset_path = '.temp_wavs/test';
estimates_base_path = 'output';

% Only evaluate the first track
evaluate_first_track_only = false;

% The stem you want to evaluate
target_stem = 'vocals';

% All possible MUSDB stems
all_stems = {'vocals', 'drums', 'bass', 'other'};

% Directory for PEASS temporary calculation files
options.destDir = 'peass_temp_output/';
if ~exist(options.destDir, 'dir')
    mkdir(options.destDir);
end

% ==========================================
% 2. EVALUATE LOOP
% ==========================================
% Get all track directories in the test set
tracks = dir(wav_dataset_path);
tracks = tracks([tracks.isdir] & ~ismember({tracks.name}, {'.', '..'}));

fprintf('Found %d tracks in the test subset.\n', length(tracks));

evaluated_count = 0;
ops_total = 0;
tps_total = 0;
ips_total = 0;
aps_total = 0;

for i = 1:length(tracks)
    track_name = tracks(i).name;
    fprintf('\nProcessing: %s\n', track_name);
    
    track_wav_dir = fullfile(wav_dataset_path, track_name);
    target_wav_path = fullfile(track_wav_dir, sprintf('%s.wav', target_stem));
    
    if ~exist(target_wav_path, 'file')
        fprintf('  [Warning] Reference WAV not found at %s. Skipping.\n', target_wav_path);
        continue;
    end
    
    % Locate your model's estimate for this specific track
    estimate_file = fullfile(estimates_base_path, sprintf('%s.stem', track_name), sprintf('%s.wav', target_stem));
    
    if ~exist(estimate_file, 'file')
        fprintf('  [Warning] Estimate not found at %s. Skipping evaluation.\n', estimate_file);
        continue;
    end
    
    % Construct the reference cell array (Target MUST be first)
    reference_files = {target_wav_path};
    
    % Add the remaining stems as interferences
    for j = 1:length(all_stems)
        stem = all_stems{j};
        if ~strcmp(stem, target_stem)
            interf_path = fullfile(track_wav_dir, sprintf('%s.wav', stem));
            reference_files{end+1} = interf_path;
        end
    end
    
    fprintf('  Computing PEASS metrics...\n');
    try
        info = audioinfo(reference_files{1});
        fs = info.SampleRate;
        total_samples = info.TotalSamples;
        
        chunk_length_sec = 10;
        chunk_samples = chunk_length_sec * fs;
        
        num_chunks = ceil(total_samples / chunk_samples);
        
        ops_sum = 0; tps_sum = 0; ips_sum = 0; aps_sum = 0;
        
        for c = 1:num_chunks
            start_idx = (c-1)*chunk_samples + 1;
            end_idx = min(c*chunk_samples, total_samples);
            
            % Read chunks
            temp_refs = cell(size(reference_files));
            for ref_idx = 1:length(reference_files)
                [y, ~] = audioread(reference_files{ref_idx}, [start_idx, end_idx]);
                temp_filename = sprintf('peass_temp_output/ref_%d.wav', ref_idx);
                audiowrite(temp_filename, y, fs);
                temp_refs{ref_idx} = temp_filename;
            end
            
            [y_est, ~] = audioread(estimate_file, [start_idx, end_idx]);
            temp_est = 'peass_temp_output/est.wav';
            audiowrite(temp_est, y_est, fs);
            
            res = PEASS_ObjectiveMeasure(temp_refs, temp_est, options);
            ops_sum = ops_sum + res.OPS;
            tps_sum = tps_sum + res.TPS;
            ips_sum = ips_sum + res.IPS;
            aps_sum = aps_sum + res.APS;
        end
        
        ops_avg = ops_sum / num_chunks;
        tps_avg = tps_sum / num_chunks;
        ips_avg = ips_sum / num_chunks;
        aps_avg = aps_sum / num_chunks;
        
        fprintf('  -> OPS (Overall):      %.2f\n', ops_avg);
        fprintf('  -> TPS (Target):       %.2f\n', tps_avg);
        fprintf('  -> IPS (Interference): %.2f\n', ips_avg);
        fprintf('  -> APS (Artifacts):    %.2f\n', aps_avg);
        
        evaluated_count = evaluated_count + 1;
        ops_total = ops_total + ops_avg;
        tps_total = tps_total + tps_avg;
        ips_total = ips_total + ips_avg;
        aps_total = aps_total + aps_avg;
        if evaluate_first_track_only && evaluated_count >= 1
            fprintf('evaluate_first_track_only is true. Evaluated 1 track. Stopping.\n');
            break;
        end
        
    catch ME
        fprintf('  [Error] Failed to evaluate %s: %s\n', track_name, ME.message);
    end
end

if evaluated_count > 0
    fprintf('\nAverage PEASS across %d evaluated track(s):\n', evaluated_count);
    fprintf('  -> OPS (Overall):      %.2f\n', ops_total / evaluated_count);
    fprintf('  -> TPS (Target):       %.2f\n', tps_total / evaluated_count);
    fprintf('  -> IPS (Interference): %.2f\n', ips_total / evaluated_count);
    fprintf('  -> APS (Artifacts):    %.2f\n', aps_total / evaluated_count);
else
    fprintf('\nNo tracks were evaluated, so no average PEASS score could be computed.\n');
end
